engine/src/main/java/org/pentaho/di/job/entries/ftpdelete/JobEntryFTPDelete.java

```diff
@@ -0,0 +1,1248 @@
+              sshclient.rm( sourceFolder + filelist[i] );
+            }
+            if ( isDetailed() ) {
+              logDetailed( "JobEntryFTPDelete.RemotefileDeleted", filelist[i] );
+            }
+            updateDeletedFiles();
+          }
+        } catch ( Exception e ) {
+          // Update errors number
+          updateErrors();
+          logError( BaseMessages.getString( PKG, "JobFTP.UnexpectedError", e.getMessage() ) );
+
+          if ( successConditionBroken ) {
+            throw new Exception( BaseMessages.getString( PKG, "JobEntryFTPDelete.SuccesConditionBroken" ) );
+          }
+        }
+      } // end for
+    } catch ( Exception e ) {
+      updateErrors();
+      logError( BaseMessages.getString( PKG, "JobEntryFTPDelete.ErrorGetting", e.getMessage() ) );
+      logError( Const.getStackTracker( e ) );
+    } finally {
+      if ( ftpclient != null && ftpclient.connected() ) {
+        try {
+          ftpclient.quit();
+          ftpclient = null;
+        } catch ( Exception e ) {
+          logError( BaseMessages.getString( PKG, "JobEntryFTPDelete.ErrorQuitting", e.getMessage() ) );
+        }
+      }
+      if ( ftpsclient != null ) {
+        try {
+          ftpsclient.disconnect();
+        } catch ( Exception e ) {
+          logError( BaseMessages.getString( PKG, "JobEntryFTPDelete.ErrorQuitting", e.getMessage() ) );
+        }
+      }
+      if ( sftpclient != null ) {
+        try {
+          sftpclient.disconnect();
+          sftpclient = null;
+        } catch ( Exception e ) {
+          logError( BaseMessages.getString( PKG, "JobEntryFTPDelete.ErrorQuitting", e.getMessage() ) );
+        }
+      }
+      if ( sshclient != null ) {
+        try {
+          sshclient.close();
+          sshclient = null;
+        } catch ( Exception e ) {
+          logError( BaseMessages.getString( PKG, "JobEntryFTPDelete.ErrorQuitting", e.getMessage() ) );
+        }
+      }
+
+      FTPClient.clearSOCKS();
+    }
+
+    result.setResult( !successConditionBroken );
+    result.setNrFilesRetrieved( NrfilesDeleted );
+    result.setNrErrors( NrErrors );
+
+    return result;
+  }
+
+  /**
+   * Checks if file is a directory
+   *
+   * @param sftpClient
+   * @param filename
+   * @return true, if filename is a directory
+   */
+  public boolean isDirectory( SFTPv3Client sftpClient, String filename ) {
+    try {
+      return sftpClient.stat( filename ).isDirectory();
+    } catch ( Exception e ) {
+      // Ignore FTP errors
+    }
+    return false;
+  }
+
+  private void SSHConnect( String realservername, String realserverpassword, int realserverport,
+    String realUsername, String realPassword, String realproxyhost, String realproxyusername,
+    String realproxypassword, int realproxyport, String realkeyFilename, String realkeyPass ) throws Exception {
+
+    /* Create a connection instance */
+
+    Connection conn = new Connection( realservername, realserverport );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *conn*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 978 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).